### DISCUSSION TITLE
Final, ClassVar and delayed initialization

### DISCUSSION THREAD

**Discussion by yaxum62** [python/typing#1271] [General]
Sometimes it is required to initialize a class variable out side of a class, e.g.
```python
class List:
   ...
List.EMPTY = List()
```

But how to annotate that class variable to be final? Few options I have considered:
```python
# EMPTY is treated as normal field instead of ClassVar
class List:
   EMPTY: Final[List]
List.EMPTY = List()

# PEP-591 does not allow this
class List
   EMPTY: Final[ClassVar[List]]
List.EMPTY = List()

# This makes sense, but not supported by standard or any checker
class List:
   EMPTY: Final[List] = Ellipsis
List.EMPTY = List()
```
Any suggestions or discussions are welcome.

**Comment by erictraut** [comment]
If you need to initialize a class variable outside of a class, then it cannot be `Final` by definition.

Why do you want to assign a class variable that is an instance of the class that contains it? That does not look like an advisable design pattern. I recommend looking at other approaches.

**Reply by yaxum62** [reply]
I want this because python class can be used as namespaces. As a namespace, it should not be picky about the type of its value. 

> That does not look like an advisable design pattern. 

If you are referring to the pattern that use class as namespace, I admit that it is not great, but not as bad as non-advisable. At lease such usage is acknowledge by several python documents [^1][^2] while seems nobody bother to fix it. On the other hand, not using namespaces to organize objects sounds anti-pattern.  Like `open()` in global namespace: convenient, but definitely not a good pattern. 


[^1]: https://docs.python.org/3/library/types.html#types.SimpleNamespace
[^2]: https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces

**Reply by jeremyn** [reply]
If your code has a class where you're not "picky about the type" of its attributes, then that's a good sign you should refine the design of that class. It's unlikely that you really mean that literally anything would be acceptable.

If you do actually mean that literally anything would be acceptable, maybe because the attribute accepts unknown user input, then you can use the `SimpleNamespace` class you linked and deal with all the messiness of unknown types when you access the attributes later.

But, in the spirit of your examples, how about this?

```python
from dataclasses import dataclass

@dataclass
class _A:
    EMPTY: list = []

A = _A()
```

(The usual warning about default mutable arguments goes here.)

I did something like that in a personal project [here](https://github.com/jeremyn/simple-ebook-manager/blob/6260bd177be6e55705c90346aacfa6b7d5a1224c/src/command.py#L204-L218) for the highlighted `CmdNames` "class", and also for `Args` in the same file. It felt pretty clumsy, but it solved the problem. Any feedback on a better approach from experts would be welcome.
